core/controllers/reader.py

```diff
@@ -122,31 +123,67 @@ def classify(exp_id, state, answer, params):
         state.interaction.id)
     normalized_answer = interaction_instance.normalize_answer(answer)
 
-    # Find the first group that satisfactorily matches the given answer. This is
-    # done by ORing (maximizing) all truth values of all rules over all answer
-    # groups. The group with the highest truth value is considered the best
-    # match.
     best_matched_answer_group = None
     best_matched_answer_group_index = len(state.interaction.answer_groups)
     best_matched_rule_spec = None
     best_matched_truth_value = 0.0
+
+    fs = fs_domain.AbstractFileSystem(fs_domain.ExplorationFileSystem(exp_id))
+    input_type = interaction_instance.answer_type
+
+    # Find the first hard rule that matches
     for (i, answer_group) in enumerate(state.interaction.answer_groups):
-        fs = fs_domain.AbstractFileSystem(
-            fs_domain.ExplorationFileSystem(exp_id))
-        input_type = interaction_instance.answer_type
-        ored_truth_value = 0.0
-        best_rule_spec = None
         for rule_spec in answer_group.rule_specs:
-            evaluated_truth_value = rule_domain.evaluate_rule(
-                rule_spec, input_type, params, normalized_answer, fs)
-            if evaluated_truth_value > ored_truth_value:
-                ored_truth_value = evaluated_truth_value
-                best_rule_spec = rule_spec
-        if ored_truth_value > best_matched_truth_value:
-            best_matched_truth_value = ored_truth_value
-            best_matched_rule_spec = best_rule_spec
-            best_matched_answer_group = answer_group
-            best_matched_answer_group_index = i
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > best_matched_truth_value:
+                    best_matched_truth_value = evaluated_truth_value
+                    best_matched_rule_spec = rule_spec
+                    best_matched_answer_group = answer_group
+                    best_matched_answer_group_index = i
+
+    # Find the maximum soft rule that matches. This is
+    # done by ORing (maximizing) all truth values of all rules over all answer
+    # groups. The group with the highest truth value is considered the best
+    # match.
+    if best_matched_answer_group is None:
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
```

Comment:
At the moment, I don't think it's so good to exactly duplicate the code between soft and hard rules. For the hard rules, it's probably fine to use the old pattern for matching a rule spec. That is, evaluate the hard rules in order and, if one of them evaluates to true (1.0), then pick that rule spec and short-circuit checking the rest (since none of them can ever be better than 1.0). This is fine for this situation, too, since the hard rules will never return anything other than 0.0 and 1.0.